chore: relocate events-example from nile-js at version 0.16.0 (#13)
TheNileDev/examples: js/events/deploy-webapp-with-pulumi/README.md

# Nile Events Example #

> This example demonstrates how synchronize your data plane and control plane in

Reviewer: ```suggestion
This example demonstrates how to synchronize (i.e., `reconcile`) your data plane and control plane in
```

real time with Nile events.

Nile doesn't prescribe any particular deployment solution, but here we'll be
using [Pulumi](https://app.pulumi.com/) to deploy objects into AWS.

> 

Reviewer: @scottydawg as a followup to [this comment](https://github.com/TheNileDev/nile-js/pull/465#discussion_r954288874), especially because we know there are users not using Pulumi, can you please add what needs to change to use something other than Pulumi?  It doesn't need to be detailed (there could be another example we develop) but at least a hint of what to do/change.

Author: That's part of what the later revision attempted to do by having a more general `Deployment` interface, and showing the `PulumiAwsDeployment` as a specialization of it. Rather than `loadPulumiStacks()`, `createStack()` and `destroyStack()`, which are all concepts specific to Pulumi, we moved to ` identifyRemoteObjects(), `createObject()` and `destroyObject()` to show that they could be more abstract. Having that interface also gives a little bit of a talking point to be able to say, "You could create a different `Deployment` implementation for Kubernetes, Terraform, etc." It might be worth bringing that one back in the next revision to make this sort of explanation easier.

I'll try to summarize that concisely in the README for now.

## Prerequisites ##

This example assumes you have:

* [An AWS account](https://aws.amazon.com/free/)
* [A Pulumi account](https://app.pulumi.com/signup) that's
  [connected to your AWS account](https://www.pulumi.com/docs/get-started/aws/begin/)

> * [The Pulumi CLI installed]()

Reviewer: ```suggestion
* [The Pulumi CLI installed](https://www.pulumi.com/docs/reference/cli/)
```

> * A Nile developer account

Reviewer: Is it worth mentioning explicitly that it needs to be a Nile developer account with a password, not passwordless with Google?

## Configure Your Control Plane ##

1. [Login to the Nile Admin Dashboard](https://nad.thenile.dev/) If you don't

>    already have one, create a workspace named "cyberdyne".

Reviewer: When I went through this workflow, I was coming off the Quickstart.  In anticipation of other users doing the same, would it be acceptable to make the workspace name consistent with the one used in the [quickstart](https://www.thenile.dev/docs/current/quick-start-ui#create-your-workspace), `clustify`?  (If yes, then I believe it requires changes in 3 places in the README.md only)

Author: Not a Terminator fan? :)

2. Create an entity type named "SkyNet". A simple schema is sufficient for
   this example:

```json
{
  "name": "SkyNet",
  "schema": {
    "type": "object",
    "properties": {
      "greeting": {
        "type": "string"
      }
    }
  }
}
```

3. Create an organization in your workspace named "sac-norad".

> 4. Create an SkyNet instance in your organization, with a simple greeting:

Reviewer: ```suggestion
4. Create a "SkyNet" instance in your organization, with a value that matches the schema defined earlier:
```

```json
{
  "greeting": "Come with me if you want to live."
}
```

## Configure Your Data Plane ##

> These instructions summarize how to [get started with Pulumi](https://www.pulumi.com/docs/get-started/aws/begin/)
> on AWS. See their docs for a more complete setup.

> Install the Pulumi CLI and set up a new Pulumi project:

Reviewer: Pulumi CLI was already installed in the pre-req section.

```bash

> [~/] $ brew install pulumi/tap/pulumi

Reviewer: ```suggestion
brew install pulumi/tap/pulumi
```

> [~/] $ mkdir nile-examples && cd nile-examples

Reviewer: ```suggestion
mkdir nile-examples && cd nile-examples
```

Author: Ah, yes. This one touches on a potentially confusing issue. I explicitly included the home directory notation here to indicate that this `nile-examples` dir is independent of this example. I configured my pulumi projects in their own directory, separate from any of my library code, not knowing what the conventional usage is. But maybe we should model this like our `iteru` repo, and have a `/pulumi` dir as a subdir of `/deploy-webapp-with-pulumi`?

Author: (If we go this ☝️ route, we'll probably want to change the subsequent examples to _not_ accept the defaults. Having a project in pulumi named "pulumi" is going to get confusing.)

> [~/] $ pulumi new aws-typescript

Reviewer: ```suggestion
pulumi new aws-typescript
```

```

> and accept the defaults:

Reviewer: Note that because I already have `nile-examples`, the defaults aren't populating.  I have to rename to get this to work seamlessly.

```
project name: (nile-examples) 
project description: (A minimal AWS TypeScript Pulumi program) 
Created project 'nile-examples'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
stack name: (dev) 
Created stack 'dev'

aws:region: The AWS region to deploy into: (us-east-1)
Saved config
```

Run `pulumi up` to ensure that you've configured Pulumi correctly. This will

> create a stack named `dev`. We won't be using this stack, but its presence

Reviewer: ```suggestion
create a new Pulumi stack named `dev`. We won't be using this stack, but its presence
```

verifies that you're ready to proceed.

## Create and Execute a Command ##

> In the `deploy-webapp-with-pulumi` directory, run `yarn install && yarn build` to

Reviewer: ```suggestion
Back up to the `deploy-webapp-with-pulumi` directory and run `yarn install && yarn build` to
```

> create the executable command binary.

Reviewer: Not sure if I'm doing something wrong but this fails:

```
[10:38:05] ~/git/examples/js/events/deploy-webapp-with-pulumi(chore/reloc-events-example) ✗: yarn install && yarn build
yarn install v1.22.19
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
success Already up-to-date.
✨  Done in 0.13s.
yarn run v1.22.19
$ shx rm -rf dist && tsc -b
src/commands/reconcile/index.ts:6:47 - error TS2307: Cannot find module './lib/pulumi' or its corresponding type declarations.

6 import { pulumiS3, PulumiAwsDeployment } from './lib/pulumi';
                                                ~~~~~~~~~~~~~~

Found 1 error.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Author: 😞  The .gitignore in the root of the examples repo ignores all `lib/` directories, so they didn't get included in the PR. I'll figure out how to fix that.

# Nile Events Example #

This example demonstrates how to synchronize (i.e., `reconcile`) your data
plane and control plane in real time with Nile events.

Nile doesn't prescribe any particular deployment solution, but here we'll be
using [Pulumi](https://app.pulumi.com/) to deploy objects into AWS. 

> If you're using another tool like Kubernetes or Terraform, you could replace

> > the `PulumiAwsDeployment` class with your own deployment implementation.

Reviewer: It would be helpful to tell the user

1. The `reconcile` code is the example here, and that this is what the user needs to provide in their own deployment
2. Where this file is, specifically in `src/commands/reconcile/`

Author: I'm not sure I follow... If they downloaded the example code, I'm assuming they know that the code is in here. I'll specify the path. (Note that it changes in the next version of this PR.)

## Configure Your Control Plane ##

> If you're not familiar with the terminology used below, be sure to read the
> [Nile Quickstart](https://www.thenile.dev/docs/current/quick-start-ui).

1. [Login to the Nile Admin Dashboard](https://nad.thenile.dev/) If you don't
   already have one, create a workspace named "clustify".
2. Create an entity type named "SkyNet". A simple schema is sufficient for
   this example:

```json
{
  "name": "SkyNet",
  "schema": {
    "type": "object",
    "properties": {
      "greeting": {
        "type": "string"
      }
    }
  }
}
```

3. Create an organization in your workspace named "sac-norad".
4. Create a "SkyNet" instance in your organization, with a value that matches 
   the schema defined earlier:

```json
{
  "greeting": "Come with me if you want to live."
}
```

## Configure Your Data Plane ##

> These instructions summarize how to [get started with Pulumi](https://www.pulumi.com/docs/get-started/aws/begin/)
> on AWS. See their docs for a more complete setup.

Set up a new Pulumi project:

```bash
mkdir pulumi && cd pulumi
pulumi new aws-typescript
```

and set the project name to "nile-examples":

```

> project name: (pulumi) nile-examples

Reviewer: 1. `nile-examples` is no longer valid
2. Strange but the project name defaults to `pulum` (notice the missing `i`) even in the folder `pulumi`.  Maybe call it `pulumi-clustify`?  (Is that reasonable what a user would call their pulumi project to match their workspace?). Or something else...

Author: I'm not sure what the pulumi conventions are. I was taking a nod from our iteru repo. Maybe @norwood or @mpokryva knows.

Author: changed to `pulumi-clustify` in latest.

project description: (A minimal AWS TypeScript Pulumi program) 
Created project 'nile-examples'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
stack name: (dev) 
Created stack 'dev'

aws:region: The AWS region to deploy into: (us-east-1)
Saved config
```

Run `pulumi up` to ensure that you've configured Pulumi correctly. This will
create a new Pulumi stack named `dev`. We won't be using this stack, but its
presence verifies that you're ready to proceed.

## Create and Execute a Command ##

Back up in the `deploy-webapp-with-pulumi` directory, run `yarn install && yarn build` 
to create the executable command binary.

> To run the binary, you'll need to pass arguments for your Nile workspace,

Reviewer: ```suggestion
To run the `reconcile` binary, you'll need to pass arguments for your Nile workspace,
```

organization, and entity type, plus your developer login. The workspace
and entity type are identified by name, but the organization requires an id.

> The organization id is not visible in the NAD yet, but can be obtained from
> the URL when you select an org. In this case:
>
> `https://nad.thenile.dev/clustify/organization/org_02qfJTCBve6bw0XlxC92CG`
>
> the organization id is `org_02qfJTCBve6bw0XlxC92CG`.

Execute the binary with:

```bash
./bin/dev reconcile --basePath https://prod.thenile.dev \
--workspace clustify \
--entity SkyNet \

> --organization YOUR_ORG_ID \

Reviewer: Just noting that I think in my earlier run this was `org`, not `organization`

Author: That was an improvement made after 0.16.0. This version requires `organization`.

--email YOUR_EMAIL \
--password YOUR_PASSWORD
```

> This will synchronize your Pulumi stack and instantiate the SkyNet instance

Reviewer: It's not clear what `instantiate the SkyNet instance` means.  Can you please elaborate?

already defined in your control plane. Check your Pulumi project dashboard to
confirm.

The command will log out your instance properties, including the `websiteUrl` 

> of the object created by the [`pulumis3` program](./src/pulumi/pulumiS3.ts):

Reviewer: ```suggestion
of the object created by the [`pulumis3` program](./src/commands/reconcile/lib/pulumi/pulumiS3.ts):
```

```bash
Outputs:

    bucketPolicy: {
        // ...redacted...
    }
    object      : {
        // ...redacted...
    }

>     websiteUrl  : "s3-website-bucket-5c7d7bc.s3-website.us-east-2.amazonaws.com"

Reviewer: Should this be redacted?

Author: No, it's a dead bucket and doesn't give away any creds. I left it in place as a sample of what they might expect to see.

Reviewer: @scottydawg makes sense.

Resources:
    + 4 created

Duration: 5s
```

> Pull up that `websiteUrl` in-browser and you'll find your provided `greeting`

Reviewer: It would help someone like me to connect the dots if you added an additional description of what this represents.  Now that a user did all the steps, recap what they've done.  i.e., something like "The `reconcile` binary got the metadata about the event of the new SkyNet instance from Nile and created the appropriate "data plane" by calling Pulumi to create the S3 resources..." (or something like that...

as well as all of your instance details.

## Add/Remove Instances ##

> In your [Nile dashboard](https://nad.thenile.dev/dev/organizations), add one or

Reviewer: ```suggestion
In the [Nile Admin Dashboard](https://nad.thenile.dev/), add one or
```

more SkyNet instances to your organization. This will trigger events that the

> command receives, and will synchronize accordingly. Deleting an instance in your

Reviewer: IIUC delete is not supported in NAD yet.  If so, maybe reword this to clarify it should be done using the API (if supported).

Author: You can delete in NAD. Select an instance and you'll see a "delete" button.

Reviewer: @scottydawg thank you for the correction, I see it now.
